### PR TITLE
Removing support for `MultiMessage` from stages

### LINKED ISSUE #1802: [FEA]: Remove `MultiMessage` and sub-classes from Morpheus
### Is this a new feature, an improvement, or a change to existing functionality?

Improvement

### How would you describe the priority of this feature request

High

### Please provide a clear description of problem this feature solves

Completely removes `MultiMessage` and its sub-classes from the Morpheus repo.

- [x] #1887
- [x] Update all the related unit tests.
- [x] Update all the related components that also use `MultiMessage`.
- [x] Update all the Morpheus examples
- [x] Update every occurrence of `MultiMessage` in the documentation to `ControlMessage`.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### LINKED ISSUE #1887: Removes the code path of `MultiMessage` from pre-process stages, post-process stages, inference stages, etc.
Previously the stages support processing both `ControlMessage` and `MultiMessage` - need to remove the code paths for `MultiMessage` from the following stages (from both C++ & Python implementation):

- [x] Pre-process stages:
  - [x]   `deserialize_stage`
  - [x]   `preprocess_base_stage`
  - [x]   `preprocess_ae_stage`
  - [x]   `preprocess_fil_stage`
  - [x]   `preprocess_nlp_stage`
  - [x]   `train_ae_stage`

- [x] Post-process stages:
  - [x]   `serialize_stage`
  - [x]   `add_scores_stage_base_stage`
  - [x]   `add_scores_stage`
  - [x]   `add_classification_stage`
  - [x]   `filter_detections_stage`
  - [x]   `generate_viz_frames_stage`
  - [x]   `ml_flow_drift_stage`
  - [x]   `time_series_stage`
  - [x]   `validation_stage`

- [x] Inference stages:
  - [x]   `auto_encoder_inference_stage`
  - [x]   `identity_inference_stage`
  - [x]   `inference_stage`
  - [x]   `pytorch_inference_stage`
  - [x]   `triton_inference_stage`

- [x] Output stages:
  - [x]   `write_to_vector_db_stage`

The related unit tests should also updated to use only `ControlMessage`.

### PR DESCRIPTION
**Pull request by yczhang-nv**
## Description

Part of finalizing `MultiMessage` deprecation. Previously the stages support processing both `ControlMessage` and `MultiMessage` - this PR removes the code paths for `MultiMessage` from the following stages (from both C++ & Python implementation):

- Pre-process stages:
  - `deserialize_stage`
  - `preprocess_base_stage`
  - `preprocess_ae_stage`
  - `preprocess_fil_stage`
  - `preprocess_nlp_stage`
  - `train_ae_stage`

- Post-process stages:
  - `serialize_stage`
  - `add_scores_stage_base_stage`
  - `add_scores_stage`
  - `add_classification_stage`
  - `filter_detections_stage`
  - `generate_viz_frames_stage`
  - `ml_flow_drift_stage`
  - `time_series_stage`
  - `validation_stage`

- Inference stages:
  - `auto_encoder_inference_stage`
  - `identity_inference_stage`
  - `inference_stage`
  - `pytorch_inference_stage`
  - `triton_inference_stage`

- Output stages:
  - `write_to_vector_db_stage`

The related unit tests are also updated to use only `ControlMessage`. 

Morpheus examples that imports stages/messages directly (rather than creating their own version) are also updated to get CI passed for this PR.

Closes #1887 

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

### CODE REVIEW

File: python/morpheus/morpheus/_lib/src/stages/filter_detections.cpp [hunk 1]
```diff
@@ -20,30 +20,32 @@
...
 #include <rmm/cuda_stream_view.hpp>               // for cuda_stream_per_thread
+#include <rmm/device_buffer.hpp>                  // for device_buffer
 #include <rmm/mr/device/per_device_resource.hpp>  // for get_current_device_resource
 
 #include <cstddef>     // for size_t
 #include <cstdint>     // for uint8_t
 #include <exception>   // for exception_ptr
 #include <functional>  // for function
-#include <memory>      // for make_shared, shared_ptr, __shared_ptr_access
+#include <memory>      // for shared_ptr, __shared_ptr_access, make_shared
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/_lib/src/stages/filter_detections.cpp:48]: Remove `__shared_ptr_access` we shouldn't be calling out implementation detail symbols.

```suggestion
#include <memory>      // for shared_ptr, make_shared
```

File: python/morpheus/morpheus/_lib/src/stages/serialize.cpp [hunk 2]
```diff
@@ -17,17 +17,16 @@
...
+#include "morpheus/messages/meta.hpp"       // for MessageMeta, SlicedMessageMeta
 #include "morpheus/objects/table_info.hpp"  // for TableInfo
 
-#include <exception>
-#include <memory>
-#include <string>
-#include <type_traits>  // for is_same_v
-#include <utility>      // for move
+#include <exception>  // for exception_ptr
+#include <memory>     // for shared_ptr, __shared_ptr_access, make_shared
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/_lib/src/stages/serialize.cpp:27]:
```suggestion
#include <memory>     // for shared_ptr, make_shared
```

File: python/morpheus/morpheus/_lib/tests/stages/test_add_classification.cpp [hunk 3]
```diff
@@ -43,7 +42,7 @@
 #include <cstdint>     // for uint8_t
 #include <filesystem>  // for operator/, path
 #include <map>         // for map
-#include <memory>      // for make_shared, allocator, __shared_ptr_access, shared_ptr
+#include <memory>      // for allocator, make_shared, __shared_ptr_access, shared_ptr
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/_lib/tests/stages/test_add_classification.cpp:45]:
```suggestion
#include <memory>      // for allocator, make_shared, shared_ptr
```

File: python/morpheus/morpheus/controllers/filter_detections_controller.py [hunk 4]
```diff
@@ -96,70 +87,60 @@ def _find_detections(self, x: MultiMessage | ControlMessage) -> typing.Union[cp.
...
-            meta = x.payload()
-            x.payload(meta.copy_ranges(true_pairs))
-            return x
-        raise TypeError(f"Unsupported message type: {type(x)}")
+        meta = msg.payload()
+        msg.payload(meta.copy_ranges(true_pairs))
+        return msg
 
-    def filter_slice(self, x: MultiMessage | ControlMessage) -> typing.List[MultiMessage] | typing.List[ControlMessage]:
+    def filter_slice(self, msg: ControlMessage) -> typing.List[ControlMessage]:
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/controllers/filter_detections_controller.py:118]:
```suggestion
    def filter_slice(self, msg: ControlMessage) -> list[ControlMessage]:
```

File: python/morpheus/morpheus/controllers/monitor_controller.py [hunk 5]
```diff
@@ -126,47 +125,47 @@ def refresh_progress(self, _):
         """
         self._progress.refresh()
 
-    def progress_sink(self, x: typing.Union[cudf.DataFrame, MultiMessage, MessageMeta, ControlMessage, typing.List]):
+    def progress_sink(self, msg: typing.Union[cudf.DataFrame, MessageMeta, ControlMessage, typing.List]):
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/controllers/monitor_controller.py:128]:
```suggestion
    def progress_sink(self, msg: cudf.DataFrame | MessageMeta | ControlMessage | list):
```

File: python/morpheus/morpheus/stages/inference/identity_inference_stage.py [hunk 6]
```diff
@@ -44,18 +44,17 @@ def __init__(self, inf_queue: ProducerConsumerQueue, c: Config):
         self._max_batch_size = c.model_max_batch_size
         self._seq_length = c.feature_length
 
-    def calc_output_dims(self, x: MultiInferenceMessage) -> typing.Tuple:
-        return (x.count, self._seq_length)
+    def calc_output_dims(self, msg: ControlMessage) -> typing.Tuple:
+        return (msg.tensors().count, self._seq_length)
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/stages/inference/identity_inference_stage.py:48]:
```suggestion
    def calc_output_dims(self, msg: ControlMessage) -> tuple[int, int]:
        return (msg.tensors().count, self._seq_length)
```

Comment (dagardner-nv) [reply python/morpheus/morpheus/stages/inference/identity_inference_stage.py:48]: Not really a part of this PR, but something I'm thinking about is if `msg.tensors().count` is a common pattern, calling `msg.tensors()` might require a bit more cost than we think, we might want to consider a `tensor_count` property.

File: docs/source/developer_guide/guides/9_control_messages.md [hunk 7]
```diff
@@ -88,12 +88,17 @@ data = cudf.DataFrame()
 msg_meta = MessageMeta(data)
```
Comment (dagardner-nv) [inline docs/source/developer_guide/guides/9_control_messages.md:88]: Update the docs above to explicitly mention that in addition to the deprecation in 24.06, in 24.10 MultiMessage was removed.

File: examples/abp_pcap_detection/abp_pcap_preprocessing.py [hunk 8]
```diff
@@ -175,22 +175,24 @@ def round_time_kernel(timestamp, rollup_time, secs):
...
         del merged_df
 
         seq_ids = cp.zeros((count, 3), dtype=cp.uint32)
-        seq_ids[:, 0] = cp.arange(x.mess_offset, x.mess_offset + count, dtype=cp.uint32)
+        seq_ids[:, 0] = cp.arange(0, count, dtype=cp.uint32)
         seq_ids[:, 2] = fea_len - 1
 
         # Create the inference memory. Keep in mind count here could be > than input count
-        memory = InferenceMemoryFIL(count=count, input__0=data, seq_ids=seq_ids)
+        memory = _messages.InferenceMemoryFIL(count=count, input__0=data, seq_ids=seq_ids)
```
Comment (dagardner-nv) [inline examples/abp_pcap_detection/abp_pcap_preprocessing.py:187]: If this is a temporary change and the various subclasses of `TensorMemory` are going to be removed in a future PR, then feel free to ignore this comment.

I'm not sure if this is an improvement here, I think in general it is bad form to import the C++ impls directly, especially in an example where we are documenting how end-users should write custom stages.

I think we should either:
1. Continue using the Python implementations of the workflow-specific subclasses of `TensorMemory`
2. Stop using them, remove them and move everything to just use `TensorMemory` directly.

File: examples/log_parsing/inference.py [hunk 9]
```diff
@@ -58,28 +57,27 @@ class TritonInferenceLogParsing(TritonInferenceWorker):
...
+    def build_output_message(self, msg: ControlMessage) -> ControlMessage:
+        seq_ids = cp.zeros((msg.tensors().count, 3), dtype=cp.uint32)
+        seq_ids[:, 0] = cp.arange(0, msg.tensors().count, dtype=cp.uint32)
+        seq_ids[:, 2] = msg.tensors().get_tensor('seq_ids')[:, 2]
 
-        memory = TensorMemory(
-            count=x.count,
+        memory = _messages.TensorMemory(
+            count=msg.tensors().count,
             tensors={
```
Comment (dagardner-nv) [inline examples/log_parsing/inference.py:67]: We should just use the Python impl of `TensorMemory` if `ControlMessage` doesn't support the the Python implementation of `TensorMemory` we should just update it to support it, mimicing what we did with the `payload_from_python_meta` method.

File: python/morpheus/morpheus/stages/preprocess/train_ae_stage.py [hunk 10]
```diff
@@ -26,10 +26,11 @@
 from morpheus.cli.register_stage import register_stage
 from morpheus.config import Config
 from morpheus.config import PipelineModes
+from morpheus.messages import ControlMessage
 from morpheus.messages.message_meta import UserMessageMeta
 from morpheus.messages.multi_ae_message import MultiAEMessage
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/stages/preprocess/train_ae_stage.py:31]: We should be removing `MultiAEMessage` as well.

**Review by dagardner-nv — CHANGES_REQUESTED** [review]

File: examples/ransomware_detection/stages/create_features.py
Comment (dagardner-nv) [inline examples/ransomware_detection/stages/create_features.py:1]: I think `examples/ransomware_detection/stages/preprocessing.py` needs to be updated as well

File: morpheus/pipeline/multi_message_stage.py
Comment (dagardner-nv) [inline morpheus/pipeline/multi_message_stage.py:1]: This looks like a merge error

File: python/morpheus/morpheus/stages/preprocess/preprocess_ae_stage.py [hunk 11]
```diff
@@ -19,16 +19,11 @@
...
 from morpheus.config import Config
 from morpheus.config import PipelineModes
 from morpheus.messages import ControlMessage
-from morpheus.messages import InferenceMemoryAE
-from morpheus.messages import MultiInferenceMessage
-from morpheus.messages import MultiMessage
-from morpheus.messages import TensorMemory as CppTensorMemory
-from morpheus.messages.multi_ae_message import MultiAEMessage
-from morpheus.stages.inference.auto_encoder_inference_stage import MultiInferenceAEMessage
 from morpheus.stages.preprocess.preprocess_base_stage import PreprocessBaseStage
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/stages/preprocess/preprocess_ae_stage.py:27]: I think the `PreprocessBaseStage` needs to be updated as well

File: python/morpheus/morpheus/_lib/include/morpheus/stages/add_scores_stage_base.hpp [hunk 12]
```diff
@@ -17,26 +17,23 @@
...
+#include <map>       // for map
+#include <memory>    // for shared_ptr
+#include <optional>  // for optional
+#include <string>    // for string
+#include <thread>    // for operator<<
 
 namespace morpheus {
 /****** Component public implementations *******************/
-/****** AddClassificationStage********************************/
+/**************AddScoresStageBase***************************/
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/_lib/include/morpheus/stages/add_scores_stage_base.hpp:36]: Good catch!

File: python/morpheus/morpheus/utils/concat_df.py [hunk 13]
```diff
@@ -12,24 +12,21 @@
...
 
-def concat_dataframes(messages: typing.List[MessageBase]) -> pd.DataFrame:
+def concat_dataframes(messages: list[ControlMessage] | list[MessageMeta]) -> pd.DataFrame:
     """
     Concatinate the DataFrame associated with the collected messages into a single Pandas DataFrame.
 
     Parameters
     ----------
-    messages : typing.List[typing.Union[MessageMeta, MultiMessage]]
+    messages : list[ControlMessage] | list[cudf.DataFrame]
```
Comment (dagardner-nv) [inline python/morpheus/morpheus/utils/concat_df.py:29]: The type hint in the function def should match the hint in the docstring.

**Review by dagardner-nv — APPROVED** [review]
Looks good, just one small docstring fix needed

**Review by mdemoret-nv — APPROVED** [review]

**Pull request merged**